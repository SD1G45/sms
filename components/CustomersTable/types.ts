import React from "react";

export interface CustomersTableProps
  extends React.HTMLAttributes<HTMLTableElement> {
      data: string[][],
      ids: string[],
      view: boolean
  }
