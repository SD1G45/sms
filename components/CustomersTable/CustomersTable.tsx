import React, {useState } from "react";
import { CustomersTableProps } from "./types";
import {
    ViewButton,
    DivContainer,
    DataEmpty,
    Pagination,
    DataTable,
    TableHeader,
    TableBody,
    Header,
    HeaderLong,
    DataLong,
    Data,
    PagePointer,
    ListData
} from "./styles";
import Link from "next/link";

function mapDataToHeaderComponent(data: string[]): JSX.Element[] {
    const mapping: JSX.Element[] = [];

    data.map((value, i) => {
        const newItem = i == 0 ?
            <HeaderLong key={i}>{value}</HeaderLong>
            :
            <Header key={i} >{value}</Header>;
        mapping.push(newItem);
    })

    return mapping;
}

function mapDataToBodyComponent(data: string[][], page: number, ids: string[], view: boolean): JSX.Element[] {
    const mapping: JSX.Element[] = [];
    const pageIndex = page - 1;


    for (let j = 0; j < 7 && pageIndex * 7 + j < data.length; j++) {
        const value = data[pageIndex * 7 + j];
        const map: JSX.Element[] = [];
        value.map((val, i) => {
            const content = view ? 
                <Data key={i}>{val}</Data>
                :
                <ListData key={i}>
                    <Link href={"/customers/view/" + ids[pageIndex * 7 + j]}>
                        <span>
                        {val}
                        {/* charAt 0 shows how many customers*/}
                        <ViewButton disabled={val.charAt(0) == '0'}>
                            View Customers
                        </ViewButton>
                        </span>
                    </Link>
                </ListData>
            const newItem = i == 0 ?
                <DataLong key={i}>{val}</DataLong>
                :
                (i == value.length - 1 ?
                    content
                    :
                    <Data key={i}>{val}</Data>
                );
            map.push(newItem);
        })
        mapping.push(<tr key={j}>{map}</tr>)
    }

    // Fill in rest of table if empty
    while (mapping.length < 7) {
        const map: JSX.Element[] = [];
        for (let i = 0; i < 8; i++) {
            map.push(<DataEmpty key={i}/>)
        }
        mapping.push(<tr key={mapping.length}>{map}</tr>);
    }
    return mapping;
}

const CustomersTable: React.FC<CustomersTableProps> = ({
    data,
    ids,
    view
}) => {
    const headers = view ? 
        ["First Name", "Last Name", "Phone #"]
        : ['Name', 'Description', 'Customers'];
    const [page, setPage] = useState(1);
    const numOfEntries = data.length;
    const numOfPages = Math.ceil(numOfEntries / 7);

    const PaginationSection = () => {
        return (
            <Pagination>
                Page {page} of {numOfPages}
                <PagePointer onClick={() => { setPage(page <= 1 ? 1 : page - 1) }}>
                    {"< "}
                </PagePointer>
                <PagePointer onClick={
                    () => { setPage(page >= numOfPages ? numOfPages : page + 1) }
                }>
                    {" >"}
                </PagePointer>
            </Pagination>
        );
    }

    return (
        <DivContainer>
            <DataTable>
                <TableHeader>
                    <tr>
                        {
                            mapDataToHeaderComponent(headers)
                        }
                    </tr>
                </TableHeader>
                <TableBody>
                    {
                        mapDataToBodyComponent(data, page, ids, view)
                    }
                </TableBody>
            </DataTable>
            {
                // pagination when needed
                numOfPages > 1 ? <PaginationSection /> : null
            }
        </DivContainer>
    )
}

export default CustomersTable;
