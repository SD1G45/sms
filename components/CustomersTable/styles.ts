import styled from "styled-components";
import Button from "../Button";

export const DivContainer = styled.div`
    padding-left: 50px;
    width: 100%;
`;

export const DataTable = styled.table`
    padding-top: 30px;
    border-spacing: 0px 0px;
    borderCollapse: collapse; 
    width: 100%;
`;

export const TableHeader = styled.thead`
    
`;

export const TableBody = styled.tbody`

`;

export const Pagination = styled.div`
    float: right;
    padding-top: 5px;
`;

export const PagePointer = styled.div `
    cursor: pointer;
    display: inline;
    font-weight: bold;
    padding-left: 10px;
`;

export const Header = styled.th`
    border-bottom: 2.75px solid #C4C4C4;
    padding-bottom: 15px;
    padding-left: 300px;
    padding-top: 15px;
    font-size: 14px;
    width: 100%;
    overflow: hidden;
    white-space: nowrap;
`;

export const HeaderLong = styled.th`
    border-bottom: 2.75px solid #C4C4C4;
    padding-bottom: 15px;
    padding-right: 140px;
    padding-top: 15px;
    font-size: 14px;
    text-align: left;
    width: 100%;
`;

export const Data = styled.td`
    border-bottom: 2.75px solid #C4C4C4;
    padding-bottom: 15px;
    padding-left: 300px;
    padding-top: 15px;
    font-size: 14px;
    font-weight: 500;
    overflow: hidden;
    white-space: nowrap;
`;

export const DataEmpty = styled.td`
    border-bottom: 2.75px solid #C4C4C4;
    padding-bottom: 60px;
    padding-right: 0px;
    font-size: 14px;
    font-weight: 500;
`;

export const DataLong = styled(Data)`
    padding-left: 0px;
    padding-right: 210px;
    font-weight: 500;
    text-align: left;
`;

export const ViewButton = styled(Button)`
    background-color: #4881F0;
    color: white;
    cursor: pointer;
    width: 50%;
    font-size: 0.75rem;
    padding: 10px 10px
`;