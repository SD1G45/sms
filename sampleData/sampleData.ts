function couponData() {
  return [
    {
      date: "12:00 AM",
      "coupons": 0
    },
    {
      date: "1:00 AM",
      "coupons": 0
    },
    {
      date: "2:00 AM",
      "coupons": 0
    },
    {
      date: "3:00 AM",
      "coupons": 0
    },
    {
      date: "4:00 AM",
      "coupons": 0
    },
    {
      date: "5:00 AM",
      "coupons": 0
    },
    {
      date: "6:00 AM",
      "coupons": 0
    },
    {
      date: "7:00 AM",
      "coupons": 0
    },
    {
      date: "8:00 AM",
      "coupons": 0
    },
    {
      date: "9:00 AM",
      "coupons": 0
    },
    {
      date: "10:00 AM",
      "coupons": 0
    },
    {
      date: "11:00 AM",
      "coupons": 3
    },
    {
      date: "12:00 PM",
      "coupons": 7
    },
    {
      date: "1:00 PM",
      "coupons": 10
    },
    {
      date: "2:00 PM",
      "coupons": 18
    },
    {
      date: "3:00 PM",
      "coupons": 20
    },
    {
      date: "4:00 PM",
      "coupons": 20
    },
    {
      date: "5:00 PM",
      "coupons": 20
    },
    {
      date: "6:00 PM",
      "coupons": 22
    },
    {
      date: "7:00 PM",
      "coupons": 23
    },
    {
      date: "8:00 PM",
      "coupons": 25
    },
    {
      date: "9:00 PM",
      "coupons": 25
    },
    {
      date: "10:00 PM",
      "coupons": 25
    },
    {
      date: "11:00 PM",
      "coupons": 25
    },
    {
      date: "11:59 PM",
      "coupons": 25
    },
  ];
}

function customersData() {
  return [
    {
      date: "12:00 AM",
      "customers": 0
    },
    {
      date: "1:00 AM",
      "customers": 0
    },
    {
      date: "2:00 AM",
      "customers": 0
    },
    {
      date: "3:00 AM",
      "customers": 0
    },
    {
      date: "4:00 AM",
      "customers": 0
    },
    {
      date: "5:00 AM",
      "customers": 0
    },
    {
      date: "6:00 AM",
      "customers": 0
    },
    {
      date: "7:00 AM",
      "customers": 0
    },
    {
      date: "8:00 AM",
      "customers": 0
    },
    {
      date: "9:00 AM",
      "customers": 0
    },
    {
      date: "10:00 AM",
      "customers": 0
    },
    {
      date: "11:00 AM",
      "customers": 3
    },
    {
      date: "12:00 PM",
      "customers": 7
    },
    {
      date: "1:00 PM",
      "customers": 10
    },
    {
      date: "2:00 PM",
      "customers": 18
    },
    {
      date: "3:00 PM",
      "customers": 20
    },
    {
      date: "4:00 PM",
      "customers": 20
    },
    {
      date: "5:00 PM",
      "customers": 20
    },
    {
      date: "6:00 PM",
      "customers": 22
    },
    {
      date: "7:00 PM",
      "customers": 23
    },
    {
      date: "8:00 PM",
      "customers": 25
    },
    {
      date: "9:00 PM",
      "customers": 25
    },
    {
      date: "10:00 PM",
      "customers": 25
    },
    {
      date: "11:00 PM",
      "customers": 25
    },
    {
      date: "11:59 PM",
      "customers": 25
    },
  ];
}

export default { couponData, customersData};