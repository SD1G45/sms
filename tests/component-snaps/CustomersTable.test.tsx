import React from "react";
import renderer from "react-test-renderer";
import "jest-styled-components";
import CustomersTable from "../../components/CustomersTable";
import sampleData from "../../sampleData/sampleData";

const data: string[][] =  [
  [
    "List 1",
    "Description of List 1",
    "69 total customers"
  ],
  [
    "List 2", 
    "Description of List 2",
    "42 total customers"
  ],
];

const ids: string[] = [
  "sample-id-for-list-1",
  "sample-id-for-list-2"
]

describe("<CustomersTable/>", () => {
  it("renders correctly", () => {
    const tree = renderer.create(
      <CustomersTable data={data} ids={ids} view={false}/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});