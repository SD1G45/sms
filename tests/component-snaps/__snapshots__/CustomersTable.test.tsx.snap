// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomersTable/> renders correctly 1`] = `
.c7 {
  width: 100%;
  padding: 12px 14px;
  background-color: #4881F0;
  border-radius: 18px;
  border: 2px solid white;
  color: white;
  cursor: pointer;
  font-size: 1rem;
  opacity: 100%;
  overflow: hidden;
  white-space: nowrap;
}

.c7:hover {
  opacity: 70%;
}

.c0 {
  padding-left: 50px;
  width: 100%;
}

.c1 {
  padding-top: 30px;
  border-spacing: 0px 0px;
  borderCollapse: collapse;
  width: 100%;
}

.c3 {
  border-bottom: 2.75px solid #C4C4C4;
  padding-bottom: 15px;
  padding-top: 15px;
  font-size: 14px;
  width: 100%;
  overflow: hidden;
  white-space: nowrap;
  text-align: center;
  padding-right: 300px;
}

.c2 {
  border-bottom: 2.75px solid #C4C4C4;
  padding-bottom: 15px;
  padding-top: 15px;
  font-size: 14px;
  text-align: left;
  width: 100%;
}

.c4 {
  border-bottom: 2.75px solid #C4C4C4;
  padding-bottom: 15px;
  padding-top: 15px;
  padding-right: 200px;
  font-size: 14px;
  font-weight: 500;
  overflow: hidden;
  white-space: nowrap;
  text-align: left;
  width: 100%;
}

.c6 {
  padding-top: 5px;
  padding-bottom: 2px;
  padding-right: 0px;
}

.c9 {
  border-bottom: 2.75px solid #C4C4C4;
  padding-bottom: 46.5px;
  padding-right: 0px;
  font-size: 14px;
  font-weight: 500;
}

.c5 {
  text-align: left;
}

.c8 {
  background-color: #4881F0;
  color: white;
  cursor: pointer;
  width: 50%;
  font-size: 0.75rem;
  padding: 10px 10px;
}

<div
  className="c0"
>
  <table
    className="c1"
  >
    <thead
      className=""
    >
      <tr>
        <th
          className="c2"
        >
          Name
        </th>
        <th
          className="c3"
        >
          Description
        </th>
        <th
          className="c3"
        >
          Customers
        </th>
      </tr>
    </thead>
    <tbody
      className=""
    >
      <tr>
        <td
          className="c4 c5"
        >
          List 1
        </td>
        <td
          className="c4"
        >
          Description of List 1
        </td>
        <td
          className="c4 c6"
        >
          <span
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            69 total customers
            <button
              className="c7 c8"
              disabled={false}
            >
              View Customers
            </button>
          </span>
        </td>
      </tr>
      <tr>
        <td
          className="c4 c5"
        >
          List 2
        </td>
        <td
          className="c4"
        >
          Description of List 2
        </td>
        <td
          className="c4 c6"
        >
          <span
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            42 total customers
            <button
              className="c7 c8"
              disabled={false}
            >
              View Customers
            </button>
          </span>
        </td>
      </tr>
      <tr>
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
      </tr>
      <tr>
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
      </tr>
      <tr>
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
      </tr>
      <tr>
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
      </tr>
      <tr>
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
        <td
          className="c9"
        />
      </tr>
    </tbody>
  </table>
  
</div>
`;
